File: Java/insertionSort.java
public class InsertionSort {

    // Method to perform Insertion Sort
    public static void insertionSort(int[] array) {
        int n = array.length; // Get the length of the array

        // Traverse through all elements starting from the second element
        for (int i = 1; i < n; i++) {
            // Current element to be inserted in the sorted part
            int key = array[i];
            // Initialize the index of the last element in the sorted part
            int j = i - 1;

            // Move elements of the sorted part that are greater than the key
            // to one position ahead of their current position
            while (j >= 0 && array[j] > key) {
                array[j + 1] = array[j]; // Shift element to the right
                j--; // Move to the previous element
            }
            // Place the key at its correct position
            array[j + 1] = key;
        }
    }

    // Method to print the array
    public static void printArray(int[] array) {
        // Iterate through the array and print each element
        for (int value : array) {
            System.out.print(value + " ");
        }
        System.out.println();
    }

    // Main method to test the sorting algorithm
    public static void main(String[] args) {
        // Sample array to be sorted
        int[] array = {64, 34, 25, 12, 22, 11, 90};

        System.out.println("Original array:");
        printArray(array); // Print the original array

        // Call the insertion sort method
        insertionSort(array);

        System.out.println("Sorted array:");
        printArray(array); // Print the sorted array
    }
}

File: Go/insertionSort.go
package main

import (
	"fmt"
)

// insertionSort sorts an array using the insertion sort algorithm.
func insertionSort(arr []int) {
	// Loop through each element starting from the second element (index 1)
	for i := 1; i < len(arr); i++ {
		key := arr[i] // The current element to be inserted
		j := i - 1    // The last index of the sorted part of the array

		// Shift elements of the sorted part to the right to make space for the key
		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j] // Shift the larger element to the right
			j--               // Move to the previous index
		}
		arr[j+1] = key // Insert the key at its correct position
	}
}

// main function to test the insertionSort
func main() {
	arr := []int{64, 34, 25, 12, 22, 11, 90} // Sample array to be sorted
	fmt.Println("Original array:", arr)

	insertionSort(arr) // Call the sorting function

	fmt.Println("Sorted array:", arr) // Print the sorted array
}
